lib/common/ipc_server.c

```diff
@@ -585,22 +594,58 @@ pcmk__ipc_prepare_iov(uint32_t request, const GString *message,
     iov[0].iov_base = header;
 
     header->version = PCMK__IPC_VERSION;
-    header->size = message->len + 1;
-    total = iov[0].iov_len + header->size;
+
+    /* We are passed an index, which is basically how many times this function
+     * has been called.  This is how we support multi-part IPC messages.  We
+     * need to convert that into an offset into the buffer that we want to start
+     * reading from.
+     *
+     * Each call to this function can send max_send_size, but this also includes
+     * the header and a null terminator character for the end of the payload.
+     * We need to subtract those out here.
+     */
+    offset = index * (max_send_size - iov[0].iov_len - 1);
+
+    /* How much of message is left to send?  This does not include the null
+     * terminator character.
+     */
+    payload_size = message->len - offset;
+
+    /* How much would be transmitted, including the header size and null
+     * terminator character for the buffer?
+     */
+    total = iov[0].iov_len + payload_size + 1;
 
     if (total >= max_send_size) {
-        crm_trace("%s", message->str);
-        crm_err("Could not transmit message; message size %" PRIu32" bytes is "
-                "larger than the maximum of %" PRIu32, header->size,
-                max_send_size);
-        rc = EMSGSIZE;
-        pcmk_free_ipc_event(iov);
-        goto done;
-    }
+        /* The entire packet is too big to fit in a single buffer.  Calculate
+         * how much of it we can send - buffer size, minus header size, minus
+         * one for the null terminator.
+         */
+        payload_size = max_send_size - iov[0].iov_len - 1;
+
+        header->size = payload_size + 1;
+
+        iov[1].iov_base = strndup(message->str + offset, payload_size);
+        if (iov[1].iov_base == NULL) {
+            rc = ENOMEM;
+            pcmk_free_ipc_event(iov);
+            goto done;
```

Comment:
I know the `done` label was already present. However, it's not really doing anything for us unless we were to
```
done:
    if ((rc != pcmk_rc_ok) && (rc != pcmk_rc_ipc_more)) {
        pcmk_free_ipc_event(iov);
    }
    return rc;
```